src/Shared/OpportunisticIntern.cs

```diff
@@ -465,6 +465,71 @@ internal StringInternTarget(string target)
             public bool ReferenceEquals(string other) => ReferenceEquals(_target, other);
         }
 
+        /// <summary>
+        /// Wrapper over a substring of a string.
+        /// </summary>
+        internal struct SubstringInternTarget : IInternable
+        {
+            /// <summary>
+            /// Stores the wrapped string.
+            /// </summary>
+            private readonly string _target;
+
+            /// <summary>
+            /// Start index of the substring within the wrapped string.
+            /// </summary>
+            private readonly int _startIndex;
+
+            /// <summary>
+            /// Constructor of the class
+            /// </summary>
+            /// <param name="target">The string to wrap.</param>
+            /// <param name="startIndex">Start index of the substring within <paramref name="target"/>.</param>
+            /// <param name="length">Length of the substring.</param>
+            internal SubstringInternTarget(string target, int startIndex, int length)
+            {
+#if DEBUG
+                if (startIndex + length > target.Length)
+                {
+                    ErrorUtilities.ThrowInternalError("wrong length");
+                }
+#endif
+                _target = target;
+                _startIndex = startIndex;
+                Length = length;
+            }
+
+            /// <summary>
+            /// Gets the length of the target substring.
+            /// </summary>
+            public int Length { get; }
+
+            /// <summary>
+            /// Gets the n character in the target substring.
+            /// </summary>
+            /// <param name="index">Index of the character to gather.</param>
+            /// <returns>The character in the position marked by index.</returns>
+            public char this[int index] => _target[index + _startIndex];
+
+            /// <summary>
+            /// Returns the target substring as a string.
+            /// </summary>
+            /// <returns>The substring.</returns>
+            public string ExpensiveConvertToString() => _target.Substring(_startIndex, Length);
+
+            /// <summary>
+            /// Compare if the target substring is equal to the given string.
+            /// </summary>
+            /// <param name="other">The string to compare with the target substring.</param>
+            /// <returns>True if the strings are equal, false otherwise.</returns>
+            public bool IsOrdinalEqualToStringOfSameLength(string other) => (String.CompareOrdinal(_target, _startIndex, other, 0, other.Length) == 0);
```

Comment:
Note to self: This conflicts with the changes in https://github.com/microsoft/msbuild/pull/5292 and has to be fixed up after the first PR, be it this one or 5292, is merged.